ci-operator/step-registry/openshift/upgrade/qe/sanity/rollback-stable/openshift-upgrade-qe-sanity-rollback-stable-chain.yaml

```diff
@@ -0,0 +1,13 @@
+chain:
+  as: openshift-upgrade-qe-sanity-rollback-stable
+  steps:
+  - ref: cucushift-upgrade-setchannel
+  - ref: cucushift-upgrade-check-not-recommended
+  - ref: cucushift-upgrade-prehealthcheck
+  - ref: cucushift-upgrade-toimage
+  - ref: cucushift-upgrade-healthcheck
+  - ref: cucushift-upgrade-rollback
```

Comment:
Could we re-use `cucushift-chainupgrade-toimage`? It looks like the only difference is `rollback` and `upgrade` sub command here, other checks almost the same.

Reply:
I dropped a mount of functions in `rollback` step, like extract_oc, admin_ack, cco related functions which are included in `chainupgrade-toimage`. They are not required in z-rollback for now. We can involve @jianlinliu @evakhoni @JianLi-RH to discuss which is better.

Reply:
good question.. 
tbh, I'm not a big fan of the current status quo of how our steps organized in terms of a large amount of code is being copy-pasted into many different scripts (which I consider an anti-pattern) and a lack of a central utils in contrast to what we used to have before prow, **_which is quite a step back_** imo. ideally, **all** the functions code (unless _definitely_ a one-time only) should reside in a central utils, and imported by each step as needed, so then we **don't have to reuse** steps unless they do exactly as their name suggests 

as for the current **far from being ideal** state of things, it is quite a dilemma, considering that a reuse will _temporary?_ save us from yet another script that require being updated at each little change to the ci, while on the other hand a reuse will complicate things where each step now be requited to fulfill many different uses at once, some of which aren't self explanatory and contradicts the step naming.

lastly I do think a _proper_ reuse of steps, as in where it simplifies things by decreasing the overall amount of different steps to maintain, instead of complicating them, is in fact beneficial, but only as long as we create those steps to be generic enough to begin with, and not trying to force a specific step into doing what it doesn't meant to do, only because of perceivable similarities

Reply:
perhaps if we break those large complex steps that have lots of _ifs_ inside because all the use-cases, into smaller, more generic, simple and clean steps, then we'll not only be able to reuse, but encouraged doing so. the only limiting factor may be the amount of steps each chain is able to contain without stretching the limitations of the platform, and the amount of work be required for such a paradigm change

Reply:
> I dropped a mount of functions in rollback step, like extract_oc, admin_ack, cco related functions which are included in chainupgrade-toimage. 

Hmm, if the existed chainupgrade step could be reused here without too much change, then `upgrade+healthcheck` and `rollback+healthcheck` step could be done in one loop, just be the same with rollback nightly job.  Or else, there will be one more health check step `cucushift-upgrade-posthealthcheck` needed.

Reply:
> perhaps if we break those large complex steps that have lots of ifs inside because all the use-cases, into smaller, more generic, simple and clean steps, then we'll not only be able to reuse, but encouraged doing so.

There is limitation in Prow that a step can't be used more than once in a workflow. But we do need to re-use steps in chain upgrade. That's why we write upgrade in one step by iterating on target images. We know that it's not that readable and manageable in long run so I've opened https://issues.redhat.com/browse/OCPQE-19479 and I hope we can collaborate with the dev team to improve it.

Reply:
> if the existed chainupgrade step could be reused here without too much change

If we re-use chainupgrade step, we need to have an ENV to identify the stable rollback test. And we need a `if...else` for the rollback to not extract oc and use `oc adm upgrade rollback` subcommand. Is it your suggestion?

Reply:
oh I wasn't aware of the once per workflow step limitation. such a pity should I say.  
well, I guess other option would be spending some research into implementing a proper Utils, so we can keep our code nice and readable, and easy to maintain in one place, instead of copy-pasting over and over again. not in the current PR scope obviously.  
perhaps it is something I can spend some time on, however considering my limited experience on the subject of prow, I'll definitely benefit from some help.

Reply:
Thanks Evgeni for your willingness to improve it. I'd suggest you comment in [OCPQE-19479](https://issues.redhat.com/browse/OCPQE-19479) and involve more folks in the dev and qe teams into discussion.

Reply:
@jianlinliu @JianLi-RH May I have your opionion on either keeping the rollback sub-command in a separate step or integrating into chainupgrade step?

Reply:
> I guess other option would be spending some research into implementing a proper Utils

Per my understanding, it's impossible to call functions outside of a step in Prow. If we want to leverage Utils model, we might need to write upgrade in a separate place. That's how ota dev does. They write upgrade in [Origin](https://github.com/openshift/origin).

Reply:
for simplicity sake, it may worth trying `source ${SHARED_DIR}/ota/utils.sh` or something similar :thinking: 
however +1 on the dev collaboration, probably that ticket is a good place for further discussion on that subject..

Reply:
Good point! But there is also limitation in the functions that `source` can execute. Agree that let's move the discussion to the jira ticket.